webrtc/README.md

```diff
@@ -0,0 +1,160 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active |                 |
+
+Authors: [@mxinden]
+
+Interest Group: [@marten-seemann]
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Protocol
+
+Scenario: Browser _A_ wants to connect to server node _B_.
+
+### Setup - Both _A_ and _B_
+
+1. [Generate a
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate).
+
+2. [Get the certificate's
+   fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+### Connection Establishment
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ create a connection with its local certificate.
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
```

Comment:
we should also mention that the same process happens on the _B_'s side (i.e. as one of the steps in `4.`  _B_ constructs _A_'s SDP offer locally based on _A_'s incoming [STUN message USERNAME attribute](https://datatracker.ietf.org/doc/html/draft-ietf-behave-rfc3489bis-02#section-10.2.6), which must contain a DTLS fingerprint.

__
Otherwise, there's no way for B to know A's DTLS fingerprint. Of course, we can disable fingerprint verification on B's side, but that would make the DTLS conn insecure. Instead, in the current implementation, I'm populating `ice-ufrag` ICE attribute with DTLS fingerprint, which is sent to the remote as a part of STUN Binding request. That way remote (_B_) learns _A_ fingerprint and performs a DTLS handshake.

Reply:
> I'm populating ice-ufrag ICE attribute with DTLS fingerprint, which is sent to the remote as a part of STUN Binding request.

This may sound like a hack, but I haven't found a better way. Comments or suggestions are welcomed of course.

Reply:
@lightsofapollo this is related to our discussion in our last meeting. Any thoughts on this?

Reply:
> To clarify, does B receive a STUN connectivity check from A and then instantiate a peerconnection and set the offer/answer SDP? Is this done using the UDPMux trait in webrtc-rs?

Yep. Here is the code: https://github.com/libp2p/rust-libp2p/pull/2622/files#diff-038d467fe95855f8fed1e3122253ffb6bcb823fd74b83273c2ace20bf3fb3d4aR365-R370

The UDPMux implementation notifies the transport about a new address. The transport then tries to "upgrade" it to webrtc, meaning instantiate a peerconnection and construct offer/answer SDP.

Reply:
This discussion moved to https://github.com/libp2p/specs/pull/412#discussion_r911746155.